misc/metrics/src/identify.rs

```diff
@@ -96,27 +110,129 @@ impl Metrics {
     }
 }
 
-impl super::Recorder<libp2p_identify::IdentifyEvent> for super::Metrics {
+impl super::Recorder<libp2p_identify::IdentifyEvent> for Metrics {
     fn record(&self, event: &libp2p_identify::IdentifyEvent) {
         match event {
             libp2p_identify::IdentifyEvent::Error { .. } => {
-                self.identify.error.inc();
+                self.error.inc();
             }
             libp2p_identify::IdentifyEvent::Pushed { .. } => {
-                self.identify.pushed.inc();
+                self.pushed.inc();
             }
-            libp2p_identify::IdentifyEvent::Received { info, .. } => {
-                self.identify.received.inc();
-                self.identify
-                    .received_info_protocols
+            libp2p_identify::IdentifyEvent::Received { peer_id, info, .. } => {
+                {
+                    let mut protocols: Vec<String> = info
+                        .protocols
+                        .iter()
+                        .filter(|p| {
```

Comment:
What is the rationale for the filtering?

Reply:
A remote node can advertise arbitrary protocol names in its _libp2p-identify_ information. Each of these protocols advertised by the remote will result in a new metric as Prometheus treats each metric _name_ and _label_ combination as a new metric.

A malicious remote node can trigger a [cardinality explosion](https://www.robustperception.io/cardinality-is-key/) through advertising a new set of arbitrary protocols on each _libp2p-identify_ exchange. To prevent this, we filter by known protocols, thus restricting the cardinality of the metric.

Does that make sense @thomaseizinger?

Long term, we could allow the user to whitelist certain custom protocols, though I think we can solve that use-case once it actually comes up.

Reply:
> A remote node can advertise arbitrary protocol names in its _libp2p-identify_ information. Each of these protocols advertised by the remote will result in a new metric as Prometheus treats each metric _name_ and _label_ combination as a new metric.
> 
> A malicious remote node can trigger a [cardinality explosion](https://www.robustperception.io/cardinality-is-key/) through advertising a new set of arbitrary protocols on each _libp2p-identify_ exchange. To prevent this, we filter by known protocols, thus restricting the cardinality of the metric.
> 
> Does that make sense @thomaseizinger?

Is there another way we can mitigate these attacks?

Can we just cap the total number of metrics to something like 1000?

This is a really subtle, non-obvious behaviour that people will only find by digging in the source code which is not ideal IMO.

> Long term, we could allow the user to whitelist certain custom protocols, though I think we can solve that use-case once it actually comes up.

I'd prefer doing this from the get-go. At least with a "loud constructor" that makes people aware of the filtering.

Reply:
> `identify::Metric::new` would need to be renamed to `identify::Metric::with_default_protocol_whitelist` or something similar.

But I envision users to never instantiate `identify::Metrics` themselves, but instead the top level `Metrics` object.

Reply:
> I see. We never seem to abstract over `Recorder` so I am wondering why we need this as a trait and it cannot just be a regular function on `Metrics`?

With `Metrics` to you refer to the top level `Metrics` or `identify::Metrics`? On the latter, yes, it could be a method.

> If it were a regular function, we could make the `record` function loud: `record_with_default_protocol_whitelist`.

So users would call `metrics.identify.record_with_default_protocol_whitelist` and provide the list for each event?

Reply:
> > I see. We never seem to abstract over `Recorder` so I am wondering why we need this as a trait and it cannot just be a regular function on `Metrics`?
> 
> With `Metrics` do you refer to the top level `Metrics` or `identify::Metrics`? On the latter, yes, it could be a method.

Ah I guess on the top-level one, we can't because Rust doesn't support method overloading. Do we necessarily need the top-level `record` function call?

If users would generally call through to `metrics.identify` and other "sub" metrics, the "loud" function call for identify would feel less out of place.

> > If it were a regular function, we could make the `record` function loud: `record_with_default_protocol_whitelist`.
> 
> So users would call `metrics.identify.record_with_default_protocol_whitelist` and provide the list for each event?

For the `default_` one, no list needs to be provided. For an alternative one, yes one would have to pass the list (probably as reference).

Reply:
> > How about, instead of a "loud constructor", exposing the number of unrecognized protocols as well @thomaseizinger?
> > Something along the lines of:
> > ```
> > libp2p_identify_protocols{protocol = "/ipfs/id/1.0.0"} 10
> > libp2p_identify_protocols{protocol = "unrecognized"} 10
> > ```
> 
> I think that would be a good addition in any case.

I added the "unrecognized" label value in https://github.com/libp2p/rust-libp2p/pull/2734/commits/70076f428b230e25e3e4e4297a85020999d56bb7 and also adjusted the `# HELP` text to:

```
            "Number of connected nodes supporting a specific protocol, with \
             \"unrecognized\" for each peer supporting one or more unrecognized \
             protocols",
```

Reply:
> > > I see. We never seem to abstract over `Recorder` so I am wondering why we need this as a trait and it cannot just be a regular function on `Metrics`?
> > 
> > 
> > With `Metrics` do you refer to the top level `Metrics` or `identify::Metrics`? On the latter, yes, it could be a method.
> 
> Ah I guess on the top-level one, we can't because Rust doesn't support method overloading. Do we necessarily need the top-level `record` function call?
> 
> If users would generally call through to `metrics.identify` and other "sub" metrics, the "loud" function call for identify would feel less out of place.

In the current setup we **do** need the top level `record` method, as some events need to be passed to multiple `Recorder` implementations. This is the case for `SwarmEvent`s today, passed both to the `Swarm` recorder and the `Identify` recorder. The latter uses the `SwarmEvent`s to do garbage collection when a peer disconnects.

I am not 100% happy with the current ergonomics of the `libp2p-metrics` interface. I don't think having the user use the lower-level `record` methods would be an improvement as the user then needs to ensure to e.g. pass `SwarmEvent` to both the `Swarm` and the `Identify` `Recorder` implementation. Otherwise they would have a memory leak.

Reply:
That's fair.

Let's just leave as is then.

Reply:
:+1: 

It bugs me that I haven't found a better abstraction so far.